packages/http/src/auth.ts

```diff
@@ -0,0 +1,130 @@
+import { Operation, Program, deepClone, deepEquals } from "@typespec/compiler";
+import { HttpStateKeys } from "./lib.js";
+import {
+  Authentication,
+  HttpAuth,
+  HttpAuthReference,
+  HttpOperation,
+  HttpService,
+  OAuth2Flow,
+  OAuth2Scope,
+  Oauth2Auth,
+} from "./types.js";
+
+export function getAuthenticationForOperation(
+  program: Program,
+  operation: Operation
+): Authentication | undefined {
+  const operationAuth = program.stateMap(HttpStateKeys.authentication).get(operation);
+  if (operationAuth === undefined && operation.interface !== undefined) {
+    const interfaceAuth = program.stateMap(HttpStateKeys.authentication).get(operation.interface);
+    return interfaceAuth;
+  }
+  return operationAuth;
+}
+
+export function resolveAuthentication(service: HttpService): {
```

Comment:
I thikn that type should have an interface and some documentation on what each property is. The array of array of tuples are a little hard to make sense of IMO.

Wondering if we keeping a similar structure to  `ServiceAuthentication`(renamed `Authentication`)  structure we have today we can have:
```ts
export interface AuthenticationReference {
  /**
   * Either one of those options can be used independently to authenticate.
   */
  options: AuthenticationOptionReference[];
}

export interface AuthenticationOptionReference {
  /**
   * For this authentication option all the given auth have to be used together.
   */
  all: HttpAuthReference[];
}

export interface HttpServiceAuthentication {
  /** All the authentication schemes used in this service. Some might only be used in certain operations. */
  schemes: HttpAuth[]; // Not sure we need `all` prefix

  /** Represent the default authentication */
  defaultAuth: AuthenticationReference;

  operationsAuth: Map<Operation, AuthenticationReference>;
}

export interface HttpAuthReference {
  scheme: HttpAuth;

  /** in the case of oauth2  */
  scopes?: string[];
}
export function resolveAuthentication(service: HttpService): HttpServiceAuthentication;
```

Reply:
also, I know there isn't much precendent for it but we are now noticing some issue with library and emitters patching things inccorectly. So making the types `readonly` for anything new added I think would be good

Reply:
> So making the types readonly for anything new added I think would be good

Yeah, lack of referential transparency is a common source of bugs

Reply:
We maybe have not be super consistent on where to put types. Everything in `types.ts` is fine or if you want to keep things that are completely tied to a function (Like options interface) next to it, i think thats fine too.